File: src/services/audit/detectors/ad/compliance/anssi/r5-segregation.ts
/**
 * ANSSI R5 - Segregation
 * Checks network and privilege segregation:
 * - Tiered admin model
 * - Service account isolation
 * - Workstation restrictions
 */

import { ADUser, ADComputer } from '../../../../../../types/ad.types';
import { Finding } from '../../../../../../types/finding.types';
import { ldapAttrToString } from '../../../../../../utils/entity-converter';

export function detectAnssiR5Segregation(
  users: ADUser[],
  computers: ADComputer[],
  _includeDetails: boolean
): Finding {
  const issues: string[] = [];

  // Check for admin accounts logging into workstations (tier violation)
  const privilegedGroups = ['Domain Admins', 'Enterprise Admins', 'Schema Admins'];
  const admins = users.filter(
    (u) => u.memberOf?.some((dn) => privilegedGroups.some((pg) => dn.includes(`CN=${pg}`)))
  );

  // Check for service accounts with interactive logon capability
  const serviceAccounts = users.filter((u) => {
    const name = (u.sAMAccountName || '').toLowerCase();
    return name.includes('svc') || name.includes('service') || name.startsWith('sa_');
  });

  const interactiveServiceAccounts = serviceAccounts.filter((u) => {
    // Check if service account doesn't have "Deny logon locally" applied
    // This is a heuristic - real check would require GPO analysis
    return u.enabled;
  });

  if (interactiveServiceAccounts.length > 0) {
    issues.push(`${interactiveServiceAccounts.length} service accounts may allow interactive logon`);
  }

  // Check for workstations in server OUs (organizational issue)
  const workstationsInServerOu = computers.filter((c) => {
    const os = ldapAttrToString(c.operatingSystem);
    const isWorkstation = /windows 10|windows 11|windows 7|windows 8/i.test(os);
    const isInServerOu = /ou=servers|ou=server|ou=datacenter/i.test(c.dn);
    return isWorkstation && isInServerOu;
  });

  if (workstationsInServerOu.length > 0) {
    issues.push(`${workstationsInServerOu.length} workstations in server OUs (tier violation)`);
  }

  // Check admin count (too many admins indicates poor segregation)
  if (admins.length > 15) {
    issues.push(`${admins.length} privileged accounts (recommend <15 for proper segregation)`);
  }

  return {
    type: 'ANSSI_R5_SEGREGATION',
    severity: 'medium',
    category: 'compliance',
    title: 'ANSSI R5 - Segregation Issues',
    description:
      'Network and privilege segregation does not meet ANSSI R5 recommendations. Implement tiered administration model.',
    count: issues.length > 0 ? 1 : 0,
    details: issues.length > 0 ? { violations: issues, framework: 'ANSSI', control: 'R5' } : undefined,
  };
}

File: internal/audit/detectors/ad/compliance/anssi/r5-segregation.go
package anssi

import (
	"context"
	"strings"

	"github.com/etcsec-com/etc-collector/internal/audit"
	"github.com/etcsec-com/etc-collector/pkg/types"
)

// R5SegregationDetector checks ANSSI R5 segregation compliance
type R5SegregationDetector struct {
	audit.BaseDetector
}

// NewR5SegregationDetector creates a new detector
func NewR5SegregationDetector() *R5SegregationDetector {
	return &R5SegregationDetector{
		BaseDetector: audit.NewBaseDetector("ANSSI_R5_SEGREGATION", audit.CategoryCompliance),
	}
}

// Detect executes the detection
func (d *R5SegregationDetector) Detect(ctx context.Context, data *audit.DetectorData) []types.Finding {
	var issues []string

	// Check for users in multiple privileged groups (no segregation)
	for _, u := range data.Users {
		if !u.Enabled() {
			continue
		}

		privilegedGroupCount := 0
		for _, memberOf := range u.MemberOf {
			memberOfLower := strings.ToLower(memberOf)
			if strings.Contains(memberOfLower, "domain admins") ||
				strings.Contains(memberOfLower, "enterprise admins") ||
				strings.Contains(memberOfLower, "schema admins") ||
				strings.Contains(memberOfLower, "account operators") ||
				strings.Contains(memberOfLower, "backup operators") {
				privilegedGroupCount++
			}
		}

		if privilegedGroupCount > 1 {
			issues = append(issues, "Users in multiple privileged groups (no segregation)")
			break
		}
	}

	// Check if standard users have admin access
	standardUsersWithAdmin := 0
	for _, u := range data.Users {
		if !u.Enabled() {
			continue
		}
		hasServiceSPN := len(u.ServicePrincipalNames) > 0
		isAdmin := u.AdminCount

		if !hasServiceSPN && isAdmin {
			standardUsersWithAdmin++
		}
	}

	if standardUsersWithAdmin > 20 {
		issues = append(issues, "Many standard users with admin privileges")
	}

	finding := types.Finding{
		Type:        d.ID(),
		Severity:    types.SeverityHigh,
		Category:    string(d.Category()),
		Title:       "ANSSI R5 - Segregation Non-Compliant",
		Description: "Privilege segregation does not meet ANSSI R5 recommendations. Separate admin roles and minimize privilege overlap.",
		Count:       0,
	}

	if len(issues) > 0 {
		finding.Count = 1
		finding.Details = map[string]interface{}{
			"violations":             issues,
			"framework":              "ANSSI",
			"control":                "R5",
			"standardUsersWithAdmin": standardUsersWithAdmin,
		}
	}

	return []types.Finding{finding}
}

func init() {
	audit.MustRegister(NewR5SegregationDetector())
}
